High-throughput technologies allow a broad survey or *a global approach*  of gene expression and provide some advantages over the study of individual genes:

- may identify **genes expressed in dramatic fashion** in some biological state
- can reveal **patterns or signiatures of gene expression** that occur in biological samples (e.g. coordinate expression of genes whose protein products are functionally related)

## High-throughput technologies pipeline

1. total RNA or mRNA is isolated from biological samples obtained from two conditions that are compared (e.g. cells from a normal and a diseased human brain).
2. RNA is converted to cDNA using reverse transcriptase, which is inherently less susceptible to proteolytic or chemical degradation than RNA.
3. cDNA can be cloned, propagated, sequenced or packaged into libraries
4. RNA may be labeled with flourescent dyes to measure trancripts on microarrays and cDNA libraries may be sequenced by RNA-seq (NGS).

- [[RNA-seq]] - raw reads are mapped and read counts are used to **infer expression levels of exons and/or transcripts**.
- [[Microarrays ]]- expression data is analyzed to **identify differentially regulated genes** (e.g. using ANOVA and scatter plots) or **clustering of genes and/or samples**.

# Overview of the process of generating high throughput gene expression data using microarrays or RNA-seq

# Workflow in RNA-seq data analysis

# From generating sequence data to FASTQ